### ISSUE TITLE
Hang during call to eglGetConfigs()

### ISSUE THREAD

**Issue by rookie1024** [kripken/emscripten#7400]
After an update to the Emscripten toolchain for C++, any call to `eglGetConfigs` causes Emscripten to hang inside the preloaded JS function for `eglGetConfigs`.

This issue is reproducible with the source code for [this project](https://github.com/rookie1024/EmscriptenTest) - specifically, it hangs on the call at line 40 of `src/test/program.cpp`

(I will try to get a simpler MWE, as I realize that project is convoluted and may not build on other systems)

Output of `em++ --version`:
```
emcc (Emscripten gcc/clang-like replacement) 1.38.14 ((unknown revision))
Copyright (C) 2014 the Emscripten authors (see AUTHORS.txt)
This is free and open source software under the MIT license.
There is NO warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.

```

Output of `clang++ -v`:
```
clang version 7.0.0 (tags/RELEASE_700/final)
Target: x86_64-pc-linux-gnu
Thread model: posix
InstalledDir: /usr/bin
Found candidate GCC installation: /usr/bin/../lib/gcc/x86_64-pc-linux-gnu/7.3.1
Found candidate GCC installation: /usr/bin/../lib/gcc/x86_64-pc-linux-gnu/8.2.1
Found candidate GCC installation: /usr/bin/../lib64/gcc/x86_64-pc-linux-gnu/7.3.1
Found candidate GCC installation: /usr/bin/../lib64/gcc/x86_64-pc-linux-gnu/8.2.1
Found candidate GCC installation: /usr/lib/gcc/x86_64-pc-linux-gnu/7.3.1
Found candidate GCC installation: /usr/lib/gcc/x86_64-pc-linux-gnu/8.2.1
Found candidate GCC installation: /usr/lib64/gcc/x86_64-pc-linux-gnu/7.3.1
Found candidate GCC installation: /usr/lib64/gcc/x86_64-pc-linux-gnu/8.2.1
Selected GCC installation: /usr/bin/../lib64/gcc/x86_64-pc-linux-gnu/8.2.1
Candidate multilib: .;@m64
Candidate multilib: 32;@m32
Selected multilib: .;@m64
Found CUDA installation: /opt/cuda, version unknown
```

**Comment by kripken** [comment]
This recent change might be the cause: 1ec888135cd450dd226b5bfbd358b1a0443a8364 there is now a loop there. We should probably limit the loop to `numConfigs` iterations (although I thought it must be properly terminated anyhow) - does that fix things for you?

**Comment by rookie1024** [comment]
Let me have a look.

In the meantime, I wrote a simple program that contains the bug described above:
```cpp
#include <EGL/egl.h>
#include <X11/Xlib.h>
#include <X11/Xutil.h>
#include <iostream>


int main() {
  Display *            dpy  = XOpenDisplay(nullptr);
  Window               root = DefaultRootWindow(dpy);
  XSetWindowAttributes swa  = {0};
  Window               win  = XCreateWindow(dpy,
                             root,
                             0,
                             0,
                             1280,
                             720,
                             0,
                             CopyFromParent,
                             InputOutput,
                             CopyFromParent,
                             CWEventMask,
                             &swa);

  XMapWindow(dpy, win);

  EGLDisplay edpy = eglGetDisplay(reinterpret_cast<EGLNativeDisplayType>(dpy));

  EGLint mjr, mnr;
  eglInitialize(edpy, &mjr, &mnr);

  EGLConfig cfg;
  EGLint    numCfg;
  eglGetConfigs(edpy, &cfg, 1, &numCfg);

  std::cout << "done" << std::endl;
}
```

The compile commands I used with this code are as follows:
```
em++ -std=c++17 -c -o main.bc main.cpp
em++ -ldl -lEGL -lpthread -lX11 -o broken.html main.bc
```

**Comment by rookie1024** [comment]
Update: the following patch fixes the hang, although I haven't tested whether or not I end up with a correct EGL config (whether or not that's actually important for WebGL):
```diff
--- bin/broken.js	2018-10-27 15:50:37.044198710 -0700
+++ _broken.js	2018-10-27 15:49:53.214196752 -0700
@@ -5921,7 +5921,7 @@
         }
 
         // read attribList.
-        for(;;) {
+        for(let i = 0; i < config_size; ++i) {
           var param = HEAP32[((attribList)>>2)];
           if (param == 0x3021 /*EGL_ALPHA_SIZE*/) {
             var alphaSize = HEAP32[(((attribList)+(4))>>2)];
```

I'll try applying this patch to the EmscriptenTest project and see if that fixes it.

**Comment by rookie1024** [comment]
Update 2: applying the above patch does completely fix the issue and restore expected behavior.

**Comment by rookie1024** [comment]
Fixed by #7401.
